src/server/scripts/Battlegrounds/SilvershardMines/battleground_silvershard_mines.cpp

```diff
@@ -0,0 +1,1011 @@
+            case SilvershardMines::PathIds::EastNorthEnd:
+                return SilvershardMines::Spells::MineCartCourierEastNorth;
+            case SilvershardMines::PathIds::NorthWestEnd:
+                return SilvershardMines::Spells::MineCartCourierNorthWest;
+            default:
+                return 0;
+        }
+    }
+
+    uint32 GetTheLongRiderAchievementSpell(uint32 pathId) const
+    {
+        TeamId controllingTeam = TEAM_NEUTRAL;
+        if (GameObject const* cz = GetControlZone())
+            controllingTeam = cz->GetControllingTeam();
+
+        if (controllingTeam == TEAM_NEUTRAL)
+            return 0;
+
+        switch (pathId)
+        {
+            case SilvershardMines::PathIds::SouthEnd:
+                return controllingTeam == TEAM_ALLIANCE ? SilvershardMines::Spells::LongRiderAchievementSouthAlliance : SilvershardMines::Spells::LongRiderAchievementSouthHorde;
+            case SilvershardMines::PathIds::EastSouthEnd:
+            case SilvershardMines::PathIds::EastNorthEnd:
+                return controllingTeam == TEAM_ALLIANCE ? SilvershardMines::Spells::LongRiderAchievementEastAlliance : SilvershardMines::Spells::LongRiderAchievementEastHorde;
+            case SilvershardMines::PathIds::NorthWestEnd:
+            case SilvershardMines::PathIds::NorthEastEnd:
+                return controllingTeam == TEAM_ALLIANCE ? SilvershardMines::Spells::LongRiderAchievementNorthAlliance : SilvershardMines::Spells::LongRiderAchievementNorthHorde;
+            default:
+                return 0;
+        }
+    }
+
+    void DoAction(int32 actionId) override
+    {
+        if (actionId != SilvershardMines::Actions::MineCartControlChanged)
+            return;
+
+        if (me->FindNearestGameObjectWithOptions(20.0f, { .StringId = SilvershardMines::StringIds::MineDepot }))
+            if (GameObject const* cz = GetControlZone())
+                _endOfTheLineTeam = cz->GetControllingTeam();
+    }
+
+    GameObject* GetControlZone() const
+    {
+        if (me->HasAura(SilvershardMines::Spells::CartControlCapturePointUnitEast))
+            return me->GetGameObject(SilvershardMines::Spells::CartControlCapturePointUnitEast);
+
+        if (me->HasAura(SilvershardMines::Spells::CartControlCapturePointUnitNorth))
+            return me->GetGameObject(SilvershardMines::Spells::CartControlCapturePointUnitNorth);
+
+        if (me->HasAura(SilvershardMines::Spells::CartControlCapturePointUnitSouth))
+            return me->GetGameObject(SilvershardMines::Spells::CartControlCapturePointUnitSouth);
+
+        return nullptr;
+    }
+
+private:
+    TaskScheduler _scheduler;
+    TeamId _endOfTheLineTeam;
+    GuidSet _theLongRidersPlayers;
+    time_t _theLongRidersSpawnTime;
+};
+
+// 115904 - Cart Cap
+class spell_bg_silvershard_mines_cart_cap final : public SpellScript
+{
+    bool Validate(SpellInfo const* spellInfo) override
+    {
+        return ValidateSpellEffect({ { spellInfo->Id, EFFECT_0 } });
```

Comment:
You don't need to validate an effect you're currently hooking, it's done internally.